node/src/WorkQueue.ts

```diff
@@ -61,7 +63,100 @@ export class WorkQueue {
+    while (true) {
+      try {
+        await db.watch(this.mainQueueKey, this.processingKey);
+
+        const isItemInProcessingKey = await db.lpos(this.processingKey, item.id);
+        const isItemInMainQueueKey = await db.lpos(this.mainQueueKey, item.id);
+        if (isItemInProcessingKey !== null || isItemInMainQueueKey !== null) {
+          console.log("Item already exists, not added", item.id);
+          await db.unwatch();
+          return false;
+        }
+        const transaction = db.multi();
+        this.addItemToPipeline(transaction, item);
+        const results = await transaction.exec();
+
+        if (!results) {
+          console.log("Transaction failed, item not added", item.id);
+          await db.unwatch();
+        }
+        console.log("Item added successfully", item.id);
+        return true
+      } catch (e) {
+        console.log("Error", e);
+      } finally {
+        await db.unwatch();
+      }
+    }
+    return false;
+  }
+
+  /**
+   * This is a dummy function used for testing!
+   * 
+   * Adds an item to the work queue only if an item with the same ID doesn't already exist with a 50 ms sleep.
+   *
+   * This method uses WATCH to add the item atomically. The db client passed must not be used by
+   * anything else while this method is running.
+   *
+   * Returns a boolean indicating if the item was added or not. The item is only not added if it
+   * already exists or if an error (other than a transaction error, which triggers a retry) occurs.
+   *
+   * @param {Redis} db The Redis Connection, this must not be used by anything else while this method is running.
+   * @param item The item that will be added, only if an item doesn't already exist with the same ID.
+   * @returns {boolean} returns false if already in queue or true if the item is successfully added.
+   */
+  async addNewItemWithSeep(db: Redis, item: Item): Promise<boolean> {
+    while (true) {
+      try {
+        await db.watch(this.mainQueueKey, this.processingKey);
+
+        const isItemInProcessingKey = await db.lpos(this.processingKey, item.id);
+        const isItemInMainQueueKey = await db.lpos(this.mainQueueKey, item.id);
+        if (isItemInProcessingKey !== null || isItemInMainQueueKey !== null) {
+          console.log("Item already exists, not added", item.id);
+          await db.unwatch();
+          return false;
+        }
+        await new Promise<void>((resolve) => setTimeout(resolve, 50));;
+        const transaction = db.multi();
+        this.addItemToPipeline(transaction, item);
+        const results = await transaction.exec();
+
+        if (!results) {
+          console.log("Transaction failed, item not added", item.id);
+          await db.unwatch();
+        }
+        console.log("Item added successfully", item.id);
+        return true
+      } catch (e) {
+        console.log("Error", e);
+      } finally {
+        await db.unwatch();
+      }
+    }
+    return false;
+  }
```

Comment:
Could you please move this to the tests directory? You'll need to pass the keys as arguments, but it shouldn't be in the main library!